### DISCUSSION TITLE
LispWorks backend - completion stops working and raises errors after loading Trivia

### DISCUSSION THREAD

**Discussion by mdbergmann** [joaotavora/sly#513] [General]
Hi.

I found some issues in the LispWorks backend in Sly, but the Slime one seems to work fine.
@joaotavora what usually is the strategy to update those backends from Slime?
Is it use the Slime one and rename swank to slynk? Or is it diffing in new stuff one by one?

Thanks,
Manfred

**Comment by joaotavora** [comment]
Use Git in both repos in the corresponding file and find the missing commits. Then cherry-pick in the SLIME -> SLY direction.  I haven't done it in a while.  It _usually_ works fine.  You may also think of investigating and describing the specific issue (though I haven't had much time lately to look at this).

**Reply by mdbergmann** [reply]
OK, I've checked. There are actually no changes to the lispworks.lisp backend.
I'll try to find a pattern for the issue I see.

**Reply by mdbergmann** [reply]
So I've narrowed it down to a package `cl-mock` that I'm using in my test system.
When I load this library then something (the auto completion) gets out of whack.
Probably not a problem of Sly. But not sure what.

![lw-sly](https://user-images.githubusercontent.com/2143978/169684392-70c66cd1-7a52-45f9-95f5-6b9107fb9283.gif)

**Reply by mdbergmann** [reply]
Apparently it seems the problematic package is trivia.

**Reply by phoe** [reply]
Do the debugger/inspector work in any way? Can you inspect the slots of the type error in question? Maybe that can shed a bit of a light.

**Reply by mdbergmann** [reply]
Under 'more' the stack trace is actually longer.
I've expanded it:

```
Error (TYPE-ERROR) printing the following condition: #<TYPE-ERROR 80100FE593>
   [Condition of type TYPE-ERROR]

Restarts:
 0: [*ABORT] Return to SLY's top level.
 1: [ABORT] Quit process.

Backtrace:
 0: [error printing frame]
      Locals:
        DATUM = TYPE-ERROR
        ARGUMENTS = #<CONS <<error printing object>> 80302A55F1>
 1: [error printing frame]
      Locals:
        SYSTEM::ESTRING = TYPE-ERROR
        SYSTEM::EARGS = #<CONS <<error printing object>> 80302A55F1>
 2: [error printing frame]
      [No Locals]
 3: [error printing frame]
      Locals:
        RESULT-TYPE = SIMPLE-STRING
        SEQUENCES = #<CONS <<error printing object>> 80501E9CC1>
 4: [error printing frame]
      Locals:
        SYSTEM::RESULT-TYPE = SIMPLE-STRING
        SYSTEM::SEQUENCES = #<CONS <<error printing object>> 80501E9CC1>
 5: (SLYNK-COMPLETION::QUALIFIED-MATCHING "QL" #<The COMMON-LISP-USER package, 744/1024 internal, 0/4 external>)
      Locals:
        PATTERN = "QL"
        HOME-PACKAGE = #<The COMMON-LISP-USER package, 744/1024 internal, 0/4 external>
        FIRST-COLON = NIL
        STARTS-WITH-COLON = NIL
        TWO-COLONS = NIL
        PACKAGE-LOCAL-NICKNAMES = NIL
        PACKAGE-LOCAL-NICKNAMES-BY-PACKAGE = #<EQL Hash Table{0} 80100FA86B>
        NICKNAMES-BY-PACKAGE = #<EQL Hash Table{49} 80100FAA63>
        DBG::|internal-flet-name-for-SORTED-NICKNAMES| = #<Closure (FLET SLYNK-COMPLETION::SORTED-NICKNAMES) subfunction of SLYNK-COMPLETION::QUALIFIED-MATCHING 80501E9D69>
        DBG::COLLECT-EXTERNAL = (("QL-CDB:CONVERT-INDEX-FILE" QL-CDB:CONVERT-INDEX-FILE (0 1) 7.1875146E-4) ..)
        DBG::COLLECT-INTERNAL = NIL
        DBG::|internal-flet-name-for-COLLECT-EXTERNAL| = #<Closure (FLET SLYNK-COMPLETION::COLLECT-EXTERNAL) subfunction of SLYNK-COMPLETION::QUALIFIED-MATCHING 8020022959>
        USE-LIST = (#<The COMMON-LISP package, 0/4 internal, 978/1024 external> #<The HARLEQUIN-COMMON-LISP package, 2/8 internal, 379/512 external> #<The LISPWORKS package, 0/4 internal, 229/256 external>)
        PACKAGE = #<The TRIVIA.LEVEL2 package, 0/256 internal, 87/256 external>
        DBG::|tail-| = (#<The TRIVIA.LEVEL1 package, 0/16 internal, 15/16 external> ..)
        SORTED-NICKNAMES = ("TRIVIA" "TRIVIA.LEVEL2")
        DBG::|do-externals-symbols-package-| = #<The TRIVIA.LEVEL2 package, 0/256 internal, 87/256 external>
        S = #<SYMBOL <<error printing object>> 80800B6D59>
        DBG::G = 49
        DBG::G = #(TRIVIA.LEVEL2:ROW-MAJOR-ARRAY* 0 0 0 0 0 TRIVIA.LEVEL2:ROW-MAJOR-ARRAY 0 0 0 TRIVIA.LEVEL2:*ARITY-CHECK-BY-TEST-CALL* TRIVIA.LEVEL2:DEFOPTIMIZER 0 0 0 0 0 0 0 TRIVIA.LEVEL2:LAMBDA-LIST 0 0 0 0 0 ..)
        DBG::G = 512
        NICKNAME = "TRIVIA"
        DBG::|tail-| = ("TRIVIA.LEVEL2")
 6: (SLYNK-COMPLETION:FLEX-COMPLETIONS "ql" "common-lisp-user" :LIMIT 300)
 7: (SYSTEM::%INVOKE :INVISIBLEP T)
 8: (SYSTEM::%EVAL (SLYNK-COMPLETION:FLEX-COMPLETIONS "ql" '"common-lisp-user"))
 9: (EVAL (SLYNK-COMPLETION:FLEX-COMPLETIONS "ql" '"common-lisp-user"))
10: (SLYNK:EVAL-FOR-EMACS (SLYNK-COMPLETION:FLEX-COMPLETIONS "ql" '"common-lisp-user") "common-lisp-user" 40)
11: ((SUBFUNCTION 1 (SUBFUNCTION 1 (SUBFUNCTION 1 SLYNK::SPAWN-WORKER-THREAD))))
12: ((SUBFUNCTION (FLET SLYNK-BACKEND:CALL-WITH-DEBUGGER-HOOK) (TOP-LEVEL-FORM 50)) ..)
13: ((SUBFUNCTION 1 SLYNK::CALL-WITH-LISTENER))
14: (SLYNK::CALL-WITH-BINDINGS ((*PACKAGE* . #<The COMMON-LISP-USER package, 744/1024 internal, 0/4 external>) (*DEFAULT-PATHNAME-DEFAULTS* . #P"") (* :TRIVIA) (**) (***) (/ (:TRIVIA)) ...) ..)
15: ((SUBFUNCTION 1 (SUBFUNCTION 1 SLYNK::SPAWN-WORKER-THREAD)))
16: (MP::PROCESS-SG-FUNCTION 0 NIL NIL)
17: [error printing frame]
```

Not sure how to further troubleshoot.
Would it make sense to do some debug log in `SLYNK-COMPLETION::QUALIFIED-MATCHING`?

**Comment by joaotavora** [comment]
Before debug log, I'd just invoke the function from the repl with those
arguments. You won't have the same dynamic state, but it might be
sufficient to reproduce it.

**Reply by mdbergmann** [reply]
Yeah, ok.
I can reproduce:

```
CL-USER> (slynk-completion::qualified-matching "foo" (find-package 'cl-user))
; Debugger entered on #<TYPE-ERROR 8010020B6B>
```

The local variables are similar. Seems to fail while searching for matches in trivia.level2 package.

**Comment by joaotavora** [comment]
That's nice. Then try to narrow it down a bit further. It's most likely a
bug in sly, but it can also be in lispworks (in which sly would probably be
patched anyway) or some very intense mischief by trivia.

**Comment by mdbergmann** [comment]
So I've narrowed it down to a section where external symbols are processed in `qualified-matching`:

```
(do-external-symbols (s package)
```

`(symbol-name s)` fails, and `s` is also not printable.

A further stack-trace of this is:

```
Backtrace:
 0: [error printing frame]
 1: [error printing frame]
 2: [error printing frame]
 3: ((SUBFUNCTION 1 (METHOD STREAM:STREAM-WRITE-STRING (SLYNK-GRAY::SLY-OUTPUT-STREAM T))))
      Locals:
        SLYNK-GRAY::START = 0
        SLYNK-GRAY::END = 5
        SLYNK-GRAY::LEN = 8000
        COUNT = 5
        SLYNK-GRAY::FREE = 5621
        STREAM = #<SLYNK-GRAY::SLY-OUTPUT-STREAM 8160283EFB>
        SLYNK-GRAY::START = 0
        SLYNK-GRAY::END = 5
        STRING = #<SIMPLE-BMP-STRING <<error printing object>> 806005BD83>
 4: ((SUBFUNCTION (FLET SLYNK-BACKEND:CALL-WITH-LOCK-HELD) (TOP-LEVEL-FORM 126)) ..)
      Locals:
        SLYNK-LISPWORKS::LOCK = #<MP:LOCK "buffer write lock" Locked once by "slynk-worker" 81603641DB>
        FUNCTION = #<Closure 1 subfunction of (METHOD STREAM:STREAM-WRITE-STRING (SLYNK-GRAY::SLY-OUTPUT-STREAM T)) 802034FFA9>
 5: [error printing frame]
 6: [error printing frame]
 7: [error printing frame]
 8: [error printing frame]
 9: [error printing frame]
10: [error printing frame]
11: [error printing frame]
12: [error printing frame]
13: (SYSTEM::SUB-FORMAT 0 17)
14: (SYSTEM::CALL-SUB-FORMAT #<SLYNK-GRAY::SLY-OUTPUT-STREAM 8160283EFB> 0 17)
15: [error printing frame]
16: (SLYNK-COMPLETION::QUALIFIED-MATCHING "QL" #<The COMMON-LISP-USER package, 474/1024 internal, 0/4 external>)
```

This `SIMPLE-BMP-STRING` looks like this:

```
#<SIMPLE-BMP-STRING 806005BD83>
--------------------
Dimensions: (5)
Element type: BMP-CHAR
Total size: 5
Adjustable: NIL
Fill pointer: NIL
Contents:
0: #<CHARACTER <<error printing object>> 3BBFF>
1: #\L
2: #\I
3: #\S
4: #\T
```

I do not know what a SIMPLE-BMP-STRING is. Probably a LispWorks string implementation?

**Reply by joaotavora** [reply]
I don't quite follow this part.  Can you say what is the symbol `s` for which `(symbol-name s)` returns an unprintable string?  Is it the symbol `SIMPLE-BMP-STRING` (of which package?).  Or is it some other symbol `x` whose name is something like `XLIST` where `X` is unprintable?

**Reply by joaotavora** [reply]
not a unicode expert, but character 3BBFF is presumably "outside any currently defined block range": https://www.unicode.org/cgi-bin/Code2Chart 

I guess SLY's completion should probably learn to deal with this and fail more gracefully.  I don't have much time to look at this, but fortunately, it seems someone who wrote a book on CL's condition system has recently participated in this discussion *cough* @phoe *cough*

**Reply by phoe** [reply]
```
#<SIMPLE-BMP-STRING 806005BD83>
--------------------
Dimensions: (5)
Element type: BMP-CHAR
Total size: 5
Adjustable: NIL
Fill pointer: NIL
Contents:
0: #<CHARACTER <<error printing object>> 3BBFF>
1: #\L
2: #\I
3: #\S
4: #\T
```

Something has gone very, very wrong over here:

```
0: #<CHARACTER <<error printing object>> 3BBFF>
```

Is it possible to have *unprintable* characters in LispWorks? Can you inspect that object or try to grab its char-code?

**Reply by phoe** [reply]
@joaotavora I don't think that it's a matter of handling it on the condition system side - I have never seen an assumption that an unprintable character is something that's allowed to exist *at all*.

Also, it seems that BMP strings in LispWorks can handle characters that are below 16 bits in length - `3BBFF` is longer than 16 bits, so it probably cannot be the code point of that character.

What kind of monster is that character object? 😮

**Reply by mdbergmann** [reply]
I've looked at what TRIVIA.LEVEL2 exports.
There is a weird symbol called: ΛLIST which probably is the issue here.

**Reply by mdbergmann** [reply]
It's a lambda.

**Reply by mdbergmann** [reply]
Or I should say should be a lambda.

**Reply by phoe** [reply]
LispWorks Personal 7.1.2 is capable of handling and printing a symbol named `ΛLIST` correctly in its listener (not in sly or slime), so either it's a regression in LW8 or loading Trivia or its dependencies breaks something much more fundamental in LispWorks. Weird stuff!

Can you bisect and check whether loading one of Trivia dependencies still causes this behavior to manifest? I wonder if you can e.g. load all of Trivia's dependencies and then try to `(print 'ΛLIST)` or something similar, and then check if loading Trivia itself breaks it even further.

**Reply by phoe** [reply]
Random question - what is the stream element type of streams that are created by sly? Is it possible that `#\Λ` is not printable because the sly streams are too narrow to contain an e.g. non-base-char?

**Reply by joaotavora** [reply]
Funny, it seems that the "greek small letter lambda is 0x03BB, so it could be related to that 0x3BBFF? Like someone left shifted some F's in there.  But at any rate the lambda that @mdbergmann posted is capital.

> I have never seen an assumption that an unprintable character is something that's allowed to exist at all.

"I don't believe in witches, but it's undeniable there are some" (spanish proverb "que las hay, las hay")

Meaning, be that as it may, but if SLY could somehow handle the situation, like discard the symbol or sth, rather than breaking completely and mysteriously, it would be better.

**Reply by joaotavora** [reply]
> Random question - what is the stream element type of streams that are created by sly? Is it possible that #\Λ is not printable because the sly streams are too narrow to contain an e.g. non-base-char?

I should be able to answer that question, but alas I can't because I'm away from my Lisp-fu at the moment.  But it's a good one.  I would guess this was all inherited from SLIME and the streams are all `UTF-8`.  That's the "coding system" at least.  But the element type is probably something different.

**Reply by phoe** [reply]
Hmm. Maybe `0x03BBFF` is LispWorks's internal character encoding and `FF` is its lowtag-like-thing denoting a character? We'd need to ask LispWorks guys.

> Meaning, be that as it may, but if SLY could somehow handle the situation, like discard the symbol or sth, rather than breaking completely and mysteriously, it would be better.

TBH I think that handling the situation in which a character or a string may not be printable is defensive programming taken to the absolute extreme, since characters and strings are defined to be *always* printable by the specification and whenever this is false, then the real bug is elsewhere, probably even on the implementation side.

My current hypothesis is:

1) the stream is unable to handle the capital lambda for whatever reason,
2) the above signals a printing error with a report that goes like `Cannot print the character #\Λ to the stream because it only accepts base-chars or something`,
3) the printing error is reported to another stream which is also unable to handle the `#\Λ` character,
4) everything breaks down except it *somehow* does not descend into stack-blowing recursion.

That would explain the `#<SIMPLE-BMP-STRING <<error printing object>> 806005BD83>` behavior I think, because the streams that are used for error reporting also cannot accept a `#\Λ` being passed into them so Sly falls back to something like `print-unreadable-object` like [here](https://github.com/joaotavora/sly/blob/5966d68727898fa6130fb6bb02208f70aa8d5ce3/slynk/slynk.lisp#L1751-L1771). I'm just curious if it's true.

**Reply by mdbergmann** [reply]
> LispWorks Personal 7.1.2 is capable of handling and printing a symbol named `ΛLIST` correctly in its listener (not in sly or slime), so either it's a regression in LW8 or loading Trivia or its dependencies breaks something much more fundamental in LispWorks. Weird stuff!
> 
> Can you bisect and check whether loading one of Trivia dependencies still causes this behavior to manifest? I wonder if you can e.g. load all of Trivia's dependencies and then try to `(print 'ΛLIST)` or something similar, and then check if loading Trivia itself breaks it even further.

Indeed, `(print 'ΛLIST)` can be printed fine in LW 8 listener.
But trying to print in it SLY breaks the connection, so probably slynk crashes, without load trivia.

**Reply by phoe** [reply]
Bingo!

What does the `*inferior-lisp*` buffer say, if you have access to it on LispWorks? Alternatively, can you spawn a slynk server on a LispWorks running with a listener and check if any sort of errors end up being printed there?

**Reply by mdbergmann** [reply]
This is the SLYNK backtrace:

```
;; Backtrace:
0: (SLYNK::SIGNAL-SLYNK-ERROR #<CONDITIONS::STREAM-ELEMENT-TYPE-ERROR 801000E283> :DONT-KNOW)
1: (COMMON-LISP:SIGNAL #<CONDITIONS::STREAM-ELEMENT-TYPE-ERROR 801000E283>)
2: [error printing frame]
3: [error printing frame]
4: [error printing frame]
5: (SLYNK-RPC::TRANSLATING-READ)
6: (SLYNK-RPC::TRANSLATING-READ)
7: (SLYNK-RPC::TRANSLATING-READ)
8: [error printing frame]
9: (SLYNK-RPC:READ-MESSAGE #<COMM:SOCKET-STREAM 80D02707B3> #<The SLYNK-IO-PACKAGE package, 4/16 internal, 0/16 external>)
10: (SLYNK::DECODE-MESSAGE #<COMM:SOCKET-STREAM 80D02707B3>)
11: (SLYNK::READ-LOOP #<SLYNK::MULTITHREADED-CONNECTION 80D027031B>)
12: (MP::PROCESS-SG-FUNCTION 0 NIL NIL)
13: [error printing frame]
;; Connection to Emacs lost. [
;;  condition: Cannot write CHARACTER U+039B into stream #<SYSTEM::STRING-OUTPUT-STREAM 8190837A63> with element-type BASE-CHAR.
;;  type: CONDITIONS::STREAM-ELEMENT-TYPE-ERROR
;;  style: :SPAWN]
```

**Reply by phoe** [reply]
On LispWorks 7.1.2 personal:

```
CL-USER 3 > (with-output-to-string (s) (print (stream-element-type s)))
BASE-CHAR 
""
```

This seems like a LispWorks conformance bug because [CLHS Macro WITH-OUTPUT-TO-STRING](http://www.lispworks.com/documentation/lw70/CLHS/Body/m_w_out_.htm#with-output-to-string) explicitly says, emphasis mine:

> element-type---a [type specifier](http://www.lispworks.com/documentation/lw70/CLHS/Body/26_glo_t.htm#type_specifier); evaluated. **The default is [character](http://www.lispworks.com/documentation/lw70/CLHS/Body/t_ch.htm#character).**

This causes the `with-output-to-string` calls below to default to `base-char`, which means that extended characters cannot be printed to the string stream.

https://github.com/joaotavora/sly/blob/5966d68727898fa6130fb6bb02208f70aa8d5ce3/slynk/slynk-rpc.lisp#L101-L145

...but then, plot twist! [LispWorks manpage for `with-output-to-string`](http://www.lispworks.com/documentation/lw60/LW/html/lw-634.htm) says, emphasis mine:

> The macro with-output-to-string behaves as specified in the ANSI Common Lisp Standard with one exception: **the default value of element-type is the value of [\*default-character-element-type\*](http://www.lispworks.com/documentation/lw60/LW/html/lw-839.htm#59860)**.
>
> Therefore for strict compliance you must call [set-default-character-element-type](http://www.lispworks.com/documentation/lw60/LW/html/lw-908.htm#30578) to set the default string type to `character`.

So slynk can probably work around this on its own by explicitly calling `set-default-character-element-type` in order to set it to `character` during slynk setup.

-------------

Also, slime has also encountered this: https://github.com/slime/slime/issues/691

**Reply by mdbergmann** [reply]
Hmm. OK.
Can this be worked around at a central place (I mean setting the default character element type)? Because this issue seems to manifest in different areas, rpc and completion, and maybe others.

**Reply by phoe** [reply]
Yes, it seems that the call to `lw:set-default-character-element-type` works globally. In particular, calling it in the LW rc file before loading slynk should work around the whole affair.

**Reply by mdbergmann** [reply]
Ah, sorry. I think you already mentioned it: `during slynk setup`.
Probably in the LW backend. I can give this a try.

**Reply by mdbergmann** [reply]
Setting it in LW startup file makes it work.
But I think this should be in SLY/SLYNK, no?

**Reply by phoe** [reply]
I think it's somewhat safe to assume that everyone who uses slynk will prefer Unicode support to 8-bit character efficiency, so, yes, IMO Slynk should set that during its setup.

**Comment by mdbergmann** [comment]
Anyway, thank you both for your help.

**Comment by joaotavora** [comment]
Away from my PC (at the 🏖️ hehe) but I've just skimmed this enthralling
discussion, and I gather a simple patch/pr to sly's lispworks backend is in
order.  If so, fantastic, and I suppose the patch would apply to slime too
(except that slime doesn't immediately seem to have user code triggering
this...)
